Make canadarm's xacro file configurable

ana-GT opened the pull request:
Hi :) . I had some issues/errors while trying to use the Canadarm demo. I fixed said issues with this PR (+ [one other PR in the demos repo](https://github.com/space-ros/demos/pull/21)). The changes are the following:

1. Missing collision meshes: Commented them out and use the .stl versions of the visual meshes instead.
2. Create a xacro file for the canadarm to replace the current urdf. The xacro file has as an argument the xyz/rpy location for the arm base, making it easier to configure the pose.
3. Create a macro for the gazebo's ros2 control interface. The macro has an argument for the type of hardware interface. The original urdf had an **effort** interface, which didn't work. Changed it to be **position** , which works fine.  

Tested with these changes (and the aforementioned demos's PR) and it works as expected.  Tested in a ROS Humble environment.

https://github.com/space-ros/simulation/assets/555379/37d77884-ada8-42cc-b4c2-7934d7b2bc05

ana-GT committed:
Fixed paths

ana-GT committed:
Set interface as an argument

Bckempa commented:
@quarkytale I tagged you for review since you're our current demo expert but if you don't have room on your plate for this we can find someone else, thanks!

quarkytale commented:
Hi @ana-GT, thanks for this contribution, really helps streamlining the demos. Question: how are you testing this, using `ros2 run rqt_joint_trajectory_controller rqt_joint_trajectory_controller`? I'm not able to control the joints, `rqt_joint_trajectory_controller` throws `rqt_qui` module not found. If that's the correct way to test, I'll recheck my installation, otherwise what am I missing?

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 -1.5708 0" xyz="0.25082 0 -0.175"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/joint_combine_link3.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

ana-GT replied:
Deleted the commented out line, per reviewer's suggestion

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 0 0" xyz="0.175 0 -0.25082"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/joint_combine_link3.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

ana-GT replied:
Deleted the commented out line, per reviewer's suggestion

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 -1.5708 0" xyz="0.25082 0 -0.175"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/joint_combine_link3.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="-1.5708 0 -1.5708" xyz="0.25082 0 -0.175"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/long_link_v2_1_curr.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 1.5708 0" xyz="7.36082 0 -0.175"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/joint_combine_link3.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 1.5708 0" xyz="-0.25082 0 -0.175"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/link_joint.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="1.57 -3.14 1.57" xyz="-7.36082 0 -0.175"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/long_link_2.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 -1.57 0" xyz="-7.36082 0 -0.175"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/joint_combine_link3.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 -3.14 0" xyz="0 0 0"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/joint_combine_link3.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+    <collision>
+      <origin rpy="0 -1.5708 0" xyz="0 0 0"/>
+      <geometry>
+        <!--<mesh filename="$(find simulation)/models/canadarm/meshes/joint_combine_link3.stl" scale="1.0 1.0 1.0" />-->
```
```suggestion
```

quarkytale commented on models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro:
```diff
@@ -0,0 +1,352 @@
+<?xml version="1.0" ?>
```
Fine with this being added but the `models/canadarm/urdf/SSRMS_Canadarm2.urdf` should not be removed as canadarm_moveit_config depends on that. Note moveit doesn't recognize `package://` paths hence they were hardcoded.

ana-GT commented:
Hi, apologies for the late reply, I was dead busy with a project. Will address the suggested changes and will reply soon on regards to the testing procedure I used (I think I created a rospy node to send motion commands, but not quite sure, I'll get back to you on that shortly)

quarkytale commented:
No worries thanks for working on this! Ideally it would be better to use the `gz_ros2_control` interface itself since we're aiming to update that in this PR.

quarkytale commented:
Update: I could control the arm using `ros2 action` command (thanks to Ezra), planning to approve this after changes are addressed.

ana-GT committed:
Merge branch 'main' into hotfix/fix_paths

ana-GT committed:
Added suggested changes: Delete commented out lines in urdf

ana-GT committed:
Deleted another commented out line that was missed in the last commit

ana-GT commented:
> Update: I could control the arm using `ros2 action` command (thanks to Ezra), planning to approve this after changes are addressed.

Happy to hear that! I pushed a commit that addressed the comments above (delete the commented out lines).

I only have one more change I'd like to make, but first wanted to check: So, the robot loads fine in Gazebo, but if you try loading it in **Rviz, the user gets an error that the meshes are not being found.** From what I can tell, the reason is that we are using:
```
<mesh filename="$(find simulation)/models...
```

If we replace it with:

```
<mesh filename="package://simulation/models...
```
**then the arm shows up fine both in Gazebo and Rviz2**. Is there any reason why not use **package** in the urdf.xacro? I think I read in the comments above that MoveIt might have trouble with it. If that is not longer the case, I would be happy to add that last change to this PR. Let me know! 
![gz_rviz](https://github.com/space-ros/simulation/assets/555379/d89bda3c-679a-48d1-9425-1aadf50b094d)

PS.- Tested this time in the **demos_ws** docker.

quarkytale commented:
Apologies for the late reply, I was travelling. I think the moveit_config is using `models/canadarm/urdf/SSRMS_Canadarm2.urdf` directly and I'm assuming you're launching RViz with the new `models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro` file you added right? If so, feel free to make the package path edit. We're reworking the moveit demo anyway.

ana-GT committed:
Changed find pkg to package://pkg so meshes are displayed in Rviz2

ana-GT committed:
Changed find pkg to package://pkg so meshes are displayed in Rviz2

ana-GT commented:
> Apologies for the late reply, I was travelling. I think the moveit_config is using `models/canadarm/urdf/SSRMS_Canadarm2.urdf` directly and I'm assuming you're launching RViz with the new `models/canadarm/urdf/SSRMS_Canadarm2.urdf.xacro` file you added right? If so, feel free to make the package path edit. We're reworking the moveit demo anyway.

I am just launching rviz and then using the **/robot_description** topic that is being published from robot_state_publisher.


In any case, I updated the xacro file with **"package://xxx"** instead of **"$(find xxx)"** and now you can also see the robot in Rviz: 
![test_canadarm](https://github.com/space-ros/simulation/assets/555379/32401c0c-bbb4-4b10-b04c-d0472b078a8a)


To test, as usual:

```ros2 launch canadarm canadarm.launch.py```

and then I just run any of the 3 services to move the arm:

```
ros2 service call /open_arm std_srvs/srv/Empty {}
ros2 service call /close_arm std_srvs/srv/Empty {}
ros2 service call /random_arm std_srvs/srv/Empty {}
```

quarkytale commented:
Thanks for iterating. This is great I could capture the same movement in RViz as in Gazebo!
![image](https://github.com/space-ros/simulation/assets/24695820/a9c5f0dc-bd1a-41ca-a900-3d22580d1c42)

quarkytale commented:
@Bckempa this is ready to be merged!

ana-GT commented:
If there is anything else needed for this PR to be merged, kindly let me know :) . Just checking as I just built a space-ros docker image and the canadarm simulation is failing because this PR is not yet merged. Thanks!

quarkytale commented:
Hey @ana-GT I think merging is blocked because commits need to be signed.

ana-GT committed:
Merge branch 'hotfix/fix_paths' of github.com:traclabs/simulation into hotfix/fix_paths

ana-GT commented:
@quarkytale  I see. I amended my last commit to be signed (didn't have my GPG keys set up yet :disappointed:  , but have it now :smiley:  ).  Would this PR require to create a new verified commit with all the info from the previous commits?

Bckempa commented:
Sorry @ana-GT, we tried to get this handled at the last tech meeting but we hit a few snags with some people out and others on work computers without their keys.

Two things:

1) To follow the project contribution procedure this all needs to be squashed into a commit ending in the issue number before being merged, see @ivanperez-keera's [contributing rules](https://github.com/space-ros/space-ros/blob/main/CONTRIBUTING.md) for details. The web UI forces a fast-forward rather than a merge after a squash so that has to be done on a dev machine. Feel free to do that yourself, otherwise someone else will when we get the permissions fixed.

2) To my knowledge there has never been discussion of requiring GPG signing on commits. @EzraBrooks is this new or an error in the repo's permissions configuration?

EzraBrooks commented:
I've disabled that protection. No idea how it got enabled.
